### PR TITLE
Feature/external cfn resource

### PR DESCRIPTION
**Pull request by temyers**

### CODE REVIEW

File: features/test_data/custom_resource_external_lambda/template.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Resources:
+  CustomResourceCube:
+    Type: Custom::Multiplier
+    Properties:
+      ServiceToken: arn:aws:lambda:ap-southeast-2:330045799684:function:cfnresponsechecker-CustomRe-MyCustomResourceLambda-1D3SYWRNVRHV8
```
Comment (temyers) [inline features/test_data/custom_resource_external_lambda/template.yaml]: AccountID - sanitize please. `123456789012`

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# [Choice] Python version: 3, 3.9, 3.8, 3.7, 3.6
+ARG VARIANT="3"
+FROM mcr.microsoft.com/vscode/devcontainers/python:0-${VARIANT}
+
+# [Option] Install Node.js
+ARG INSTALL_NODE="true"
+ARG NODE_VERSION="lts/*"
+RUN if [ "${INSTALL_NODE}" = "true" ]; then su vscode -c "umask 0002 && . /usr/local/share/nvm/nvm.sh && nvm install ${NODE_VERSION} 2>&1"; fi
+
+# [Optional] If your pip requirements rarely change, uncomment this section to add them to the image.
```
Comment (kallangerard) [inline .devcontainer/Dockerfile:12]: If this is going to be run from a devcontainer, consider dropping Pipefile and using a requirements.txt instead?

Comment (temyers) [reply .devcontainer/Dockerfile:12]: I wondered about that myself.

File: cfnresponsechecker/assume_role.py [hunk 3]
```diff
@@ -13,19 +13,19 @@ def __init__(self, role):
...
-        client = boto3.client(
-            "cloudformation",
-            config=config,
-            aws_access_key_id=self.credentials["AccessKeyId"],
-            aws_secret_access_key=self.credentials["SecretAccessKey"],
-            aws_session_token=self.credentials["SessionToken"],
-        )
-        stack = Stacks(client)
-        return stack.get_problem_stacks()
+    # def get_problem_stacks(self, region):
```
Comment (temyers) [inline cfnresponsechecker/assume_role.py:16]: remove dead code (can be subsequent PR)

File: cfnresponsechecker/assume_role.py [hunk 4]
```diff
@@ -34,16 +34,33 @@ def test_stack(self, region, stack_id):
...
             aws_session_token=self.credentials["SessionToken"],
         )
-        stack = Stacks(client)
-        return stack.test_stack(stack_id)
+        # stack = Stacks(client)
+        # return stack.get_problem_report()
+        return client
+
+
+    # def test_stack(self, region, stack_id):
```
Comment (temyers) [inline cfnresponsechecker/assume_role.py:42]: remove dead code (can be subsequent PR)

File: cfnresponsechecker/stacks.py [hunk 5]
```diff
@@ -44,43 +65,266 @@ def _get_stack_resources(self, stack_id):
...
+        """
+        Adapter: query AWS for active Stacks
+        """
+        stack_summaries = self._adapt_stack_summaries_from_aws()
+        if stack_summaries == None:
             return
+        stack_ids = self._get_stack_ids(stack_summaries)
+        return stack_ids
+
+    # def test_functions(self, stack_info):
```
Comment (temyers) [inline cfnresponsechecker/stacks.py:121]: remove dead code (can be subsequent PR)

File: cfnresponsechecker/stacks.py [hunk 6]
```diff
@@ -44,43 +65,266 @@ def _get_stack_resources(self, stack_id):
...
+        if len(function_report_row["functions"]) == 0:
+            return False
+
+        stack_id = function_report_row["stack"]
+        stack_resources = [stack_info["resources"] for stack_info in stack_details if stack_info["stackId"] == stack_id][0]
+        if stack_resources == None:
+            logging.debug(f"No resources found for {stack_id}")
+            return False
+
+        # custom_resources = [
```
Comment (temyers) [inline cfnresponsechecker/stacks.py:285]: remove dead code (can be subsequent PR)

File: cfnresponsechecker/stacks.py [hunk 7]
```diff
@@ -44,43 +65,266 @@ def _get_stack_resources(self, stack_id):
...
+        old_stack_resources = [
+            resource
+            for resource in stack_resources
+            if self._out_of_date_lambda_function(resource)]
+
+        return len(old_stack_resources) > 0
+
+    def _stacks_with_out_of_date_lambdas(self, function_report, stack_details):
+        return [stack["stack"] for stack in function_report if self._stack_has_out_of_date_lambdas(stack, stack_details)]
+        # stack_resources = stack_info["resources"]
```
Comment (temyers) [inline cfnresponsechecker/stacks.py:302]: remove dead code (can be subsequent PR)

File: cross_accounts.py [hunk 8]
```diff
@@ -11,10 +13,13 @@ def main(regions, role_name, accounts, clean_print=False):
         for region in regions:
             if not clean_print:
                 print(f"Region {region}")
-            problem_stacks = role.get_problem_stacks(region)
-            if problem_stacks:
-                for stack in problem_stacks:
-                    print(stack)
+            #TODO - clean up dead code
```
Comment (temyers) [inline cross_accounts.py:16]: remove dead code (can be subsequent PR)

File: features/test_data/data-generation/list_stacks.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+from botocore.config import Config
+
+def main():
+
+    config = Config(region_name='ap-southeast-2')
+    client = boto3.client("cloudformation", config=config)
+    result = paginator(client, "list_stacks", "StackSummaries", {"StackStatusFilter": [
+        'CREATE_COMPLETE','UPDATE_COMPLETE', 'UPDATE_ROLLBACK_FAILED','UPDATE_ROLLBACK_COMPLETE','IMPORT_ROLLBACK_COMPLETE', 'DELETE_FAILED'
+    ]})
+    # result = paginator(client, "list_stacks", "StackSummaries")
```
Comment (temyers) [inline features/test_data/data-generation/list_stacks.py:17]: dead code - remove.  Can be subsequent PR

File: single_account.py [hunk 10]
```diff
@@ -1,17 +1,23 @@
...
 def main(regions, role_arn, clean_print=False):
     role = Roles(role_arn)
     for region in regions:
         if not clean_print:
             print(f"Region {region}")
-        problem_stacks = role.get_problem_stacks(region)
-        if problem_stacks:
-            for stack in problem_stacks:
-                print(stack)
+        # TODO - refactor to use whole report - create a ProblemReport class
```
Comment (temyers) [inline single_account.py:14]: dead code - remove.  Can be subsequent PR

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (temyers) [reply .devcontainer/Dockerfile:12]: this can be subsequent improvement

**Review by temyers — COMMENTED** [review]
Almost good enough to merge (one main comment).
Cleaning up dead code, can be done as a subsequent PR.

**Comment by temyers** [comment]
LGTM - merging.

**Pull request merged**